cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -266,6 +266,52 @@ void EnsureLookupTablesFilled() {}
 
 #endif  // ARROW_WITH_UTF8PROC
 
+template <typename Type>
+struct AsciiReverse : StringTransform<Type, AsciiReverse<Type>> {
+  using Base = StringTransform<Type, AsciiReverse<Type>>;
+  using offset_type = typename Base::offset_type;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
+    uint8_t utf8_char_found = 0;
+    for (offset_type i = 0; i < input_string_ncodeunits; i++) {
+      // if a utf8 char is found, report to utf8_char_found
+      utf8_char_found |= input[i] & 0x80;
```

Comment:
IMO `ascii_reverse` is slightly different from the other `ascii_*` routines. If there are non-ascii chars present, they can bypass them (ex: lower, upper, etc), but in reverse, we can't do that. Therefore, I added a flag `utf8_char_found` inside the loop, and if any utf8 char is found, the `ascii_reverse` would throw an invalid status. 
I am not sure, if this check is needed or not. If we can assume that the user passes an 'ascii-only' input, we can simply user `std::reverse_copy`. So, I'd like to get a second opinion on that.

Reply:
IMO if the check is relatively cheap then it's good to keep. If someone is using the ASCII kernel explicitly, presumably either
1) they want an assurance their data is ASCII
2) they know data is ASCII and assume it's faster to reverse knowing this

Reply:
I'm ok with the check.

Reply:
Consider the case where you have a long string, would it be better to have the check inside the loop so that it stops as soon as an invalid character is found?  Perhaps using `ARROW_PREDICT_FALSE`. I understand a conditional check in a loop may break any compiler vectorization, so I would suggest verifying if this is the case.

Reply:
We don't really care about making error cases fast.

Reply:
If the ASCII version checks for invalid characters, does this means that the UTF8 version should also validate its input? Not sure if this is the case for all the other string kernels, but I think having a consistent behavior is desired.

Reply:
The ASCII version would produce invalid output for valid (but non-ASCII) input, which is why it checks for ASCII validity. Invalid input is a different situation which we don't need to check for.

Reply:
Thanks for the clarification.